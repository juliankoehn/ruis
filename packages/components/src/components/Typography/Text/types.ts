import React from 'react';
import { TypeAttributes } from '../../@types/common';
import { SerializedStyles } from '@emotion/core';

export type TextProps = {
  /**
   * Size of the text style
   * Can be: xs, sm, md, lg
   */
  size?: TypeAttributes.Size;
  fontFamily?: TypeAttributes.FontFamily;
  children: React.ReactNode;
  css?: SerializedStyles;
  className?: string;
};
