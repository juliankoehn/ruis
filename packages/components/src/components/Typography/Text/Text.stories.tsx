import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import Text from './Text';
import { TypeAttributes } from '../../@types/common';

export default {
  title: 'Typography/Text',
  decorators: [withKnobs],
};

const sizes: TypeAttributes.Size[] = ['xs', 'sm', 'md', 'lg'];
const dummyText = 'A red flair silhouetted the jagged edge of a wing.';

export const Default = () => {
  return (
    <div>
      {sizes.map((size) => (
        <div>
          <Text size={size}>
            {size}: {dummyText}
          </Text>
        </div>
      ))}
    </div>
  );
};
