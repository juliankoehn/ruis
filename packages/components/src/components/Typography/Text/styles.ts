import { css } from '@emotion/core';
import { TypeAttributes } from '../../@types/common';
import tokens from '@ruids/tokens';

export const textSizes = {
  lg: {
    fontSize: tokens.textXl,
    fontWeight: tokens.fontNormal,
    lineHeight: tokens.leadingNormal,
    letterSpacing: tokens.trackingTighter,
    marginTop: tokens.spacing6,
  },
  md: {
    fontSize: tokens.textBase,
    fontWeight: tokens.fontNormal,
    lineHeight: tokens.leading5,
    letterSpacing: tokens.trackingTight,
    marginTop: tokens.spacing4,
  },
  sm: {
    fontSize: tokens.textSm,
    fontWeight: tokens.fontNormal,
    lineHeight: tokens.leading5,
    letterSpacing: tokens.trackingTight,
    marginTop: tokens.spacing3,
  },
  xs: {
    fontSize: tokens.textXs,
    fontWeight: tokens.fontNormal,
    lineHeight: tokens.leadingNone,
    letterSpacing: tokens.trackingNormal,
    marginTop: tokens.spacing3,
  },
};

export const getTextStyle = (
  size: TypeAttributes.Size,
  font: TypeAttributes.FontFamily,
) => {
  const fontFamily = font === 'sans' ? tokens.fontSans : tokens.fontMono;
  switch (size) {
    case 'lg': {
      return css({
        ...textSizes.lg,
        fontFamily: fontFamily,
        color: tokens.colorTextDark,
        '-webkit-font-smoothing': 'auto',
        '-moz-osx-font-smoothing': 'auto',
      });
    }
    case 'md': {
      return css({
        ...textSizes.md,
        fontFamily: fontFamily,
        color: tokens.colorTextDark,
        '-webkit-font-smoothing': 'auto',
        '-moz-osx-font-smoothing': 'auto',
      });
    }
    case 'sm': {
      return css({
        ...textSizes.sm,
        fontFamily: fontFamily,
        color: tokens.colorTextDark,
        '-webkit-font-smoothing': 'auto',
        '-moz-osx-font-smoothing': 'auto',
      });
    }
    case 'xs': {
      return css({
        ...textSizes.xs,
        fontFamily: fontFamily,
        color: tokens.colorTextDark,
        '-webkit-font-smoothing': 'auto',
        '-moz-osx-font-smoothing': 'auto',
      });
    }
  }
};
