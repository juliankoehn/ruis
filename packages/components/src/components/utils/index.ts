export { canUseDOM } from './canUseDom';
export { lightenColor } from './color.lighten';
export { axe } from './axe';

export { debounce } from './debounce';
export { getScrollbarSize } from './getScrollbarSize';
export { ownerDocument } from './ownerDocument';
export { ownerWindow } from './ownerWindow';
