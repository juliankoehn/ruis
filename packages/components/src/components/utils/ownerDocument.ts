export const ownerDocument = (node: Node): Document => {
  return (node && node.ownerDocument) || document;
};
