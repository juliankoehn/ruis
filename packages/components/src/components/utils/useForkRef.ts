import { useMemo, MutableRefObject } from 'react';

export type HTMLElementOrNull = HTMLElement | null;
export type CallbackRef = (node: HTMLElement | null) => any;
export type AnyRef = CallbackRef | MutableRefObject<HTMLElementOrNull>;

function setRef(ref: AnyRef, value: HTMLElementOrNull) {
  if (typeof ref === 'function') {
    ref(value);
  } else if (ref) {
    ref.current = value;
  }
}

function useForkRef(refA: AnyRef, refB: AnyRef): CallbackRef | null {
  /**
   * This will create a new function if the ref props change and are defined.
   * This means react will call the old forkRef with `null` and the new forkRef
   * with the ref. Cleanup naturally emerges from this behavior
   */
  return useMemo(() => {
    if (refA == null && refB == null) {
      return null;
    }
    return (refValue: HTMLElementOrNull) => {
      setRef(refA, refValue);
      setRef(refB, refValue);
    };
  }, [refA, refB]);
}

export { useForkRef };
