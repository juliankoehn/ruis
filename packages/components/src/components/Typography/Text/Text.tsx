import React from 'react';
import { TextProps } from './types';
import { getTextStyle } from './styles';

const Text = React.forwardRef(
  (props: TextProps, ref: React.Ref<HTMLSpanElement>) => {
    const { className, css, size = 'sm', fontFamily = 'sans', ...rest } = props;

    const textStyle = getTextStyle(size, fontFamily);

    return (
      <span ref={ref} className={className} css={[textStyle, css]} {...rest} />
    );
  },
);

export default Text;
