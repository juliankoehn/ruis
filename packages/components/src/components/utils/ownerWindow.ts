import { ownerDocument } from './ownerDocument';

export const ownerWindow = (node: Node): Window => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};
