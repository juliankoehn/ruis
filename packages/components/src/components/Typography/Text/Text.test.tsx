import React from 'react';
import Text from './Text';
import { render } from '@testing-library/react';
import { axe } from '../../utils';

describe('Text', () => {
  it('renders the component', () => {
    const { container } = render(<Text>Text</Text>);

    expect(container).toMatchSnapshot();
  });

  it('has no a11y issues', async () => {
    const { container } = render(<Text>Text</Text>);

    expect(await axe(container)).toHaveNoViolations();
  });
});
