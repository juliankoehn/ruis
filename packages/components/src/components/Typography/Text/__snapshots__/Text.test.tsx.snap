// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders the component 1`] = `
<div>
  <span
    class="css-7kc2ge-Text"
  >
    Text
  </span>
</div>
`;
